Replace uses of the word "accessor" in diagnostics with user-facing terminology

iMostfa opened the pull request:
<!-- What's in this pull request? -->
Hello team, 

In this PR i worked on replacing the word accessor in diagnostics with more user-facing terminologies like `setter, getter, didSet Observer, and members` based on the context of the message.

in some messages i didn't need to pass `DescriptiveDeclKind` instead i just changed the text copy itself.

i also updated tests, so you might find it easier to check my changes this way.

Please let me know if there's something i should've done in a better way, and request changes if needed. !


**i can squash my commits after reviewing getting the PR Reviewed, just to make it easier to be checked commit by commit**

<!--
If this pull request resolves any GitHub issues, link them.
For information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->
Resolves #55887

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

iMostfa committed:
replace accessor to type name

iMostfa committed:
replace accessor in various diagnostics messages, delete unused one

iMostfa committed:
update final_not_on_accessors

iMostfa committed:
change change_to_mutating to use setter/getter

iMostfa committed:
update objc_init_accessor

iMostfa committed:
update tests

iMostfa commented:
Hi @kavon 
would appreciate your review & feedback here

iMostfa commented:
Hi @xedin would appreciate your review here when possible 🙏🏻

xedin reviewed (changes requested):
Sorry for the delay. Overall looks good, thank you! 

Few notes:

- `init accessors` is a feature, they are not really "functions" nor "members", so let's leave all of the uses of `init accessor` be for now.
- Removing information such as "accessor" doesn't always help because the diagnostic is applicable to a particular situation.

See more comments inline.

kavon commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7204,7 +7201,7 @@ ERROR(property_wrapper_computed, none,
       ())
 ERROR(property_wrapper_effectful,none,
-    "property wrappers currently cannot define an 'async' or 'throws' accessor",
+    "property wrappers currently cannot define an 'async' or 'throws'",
```
Maybe something like `property wrappers cannot be applied to properties that are 'async' or 'throws'` ?

iMostfa replied:
done, applied @kavon  suggestion

AnthonyLatsis replied:
> what it's trying to say is that property wrappers cannot be applied to `async` or `throwing` properties?

Actually, it is referring to `wrappedValue`:

https://github.com/swiftlang/swift/blob/3ccbcfe8cd40ded2d261b4011e1cc2e195361389/test/decl/var/effectful_property_wrapper.swift#L7-L14

Still, I don’t think we define what an `throws` or throwing property is either.

xedin replied:
Hm, how about - `"'wrappedValue' property cannot currently define an 'async' or 'throws' %0", DescriptiveDeclKind`?

iMostfa replied:
thanks @AnthonyLatsis i missed the context (as i'm editing a lot of them) 
i think @xedin new suggestion is better in this case, 1f19724a0c8

which results in the new message: 
`'wrappedValue' property cannot currently define an 'async' or 'throws' getter`

iMostfa commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7705,7 +7701,7 @@ NOTE(macro_expand_circular_reference_unnamed_through, none,
 
 ERROR(accessor_macro_not_single_var, none,
-      "accessor macro %0 can only apply to a single variable", (DeclName))
+      "macro %0 can only apply to a single variable", (DeclName))
```
done, reverted to original one: 
`"accessor macro %0 can only apply to a single variable"`

iMostfa commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7656,14 +7653,13 @@ ERROR(invalid_macro_role_for_macro_syntax,none,
 ERROR(macro_nonobserving_accessor_missing_from_expansion,none,
-      "expansion of macro %0 did not produce a non-observing accessor "
-      "(such as 'get') as expected",
+      "expansion of macro %0 did not produce a getter or a setter as expected",
```
reverted to the original one, as i don't see a way to produce a better message (aka specifying the exact missing accessor)

iMostfa commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7636,7 +7633,7 @@ ERROR(invalid_decl_in_macro_expansion,none,
       (DescriptiveDeclKind))
 ERROR(let_accessor_expansion,none,
-      "cannot expand accessors on variable declared with 'let'",
+      "cannot expand macros on variable declared with 'let'",
```
done, 
now it's
`"cannot expand accessor macro on variable declared with 'let'"`

iMostfa committed:
reword some diagnostics

iMostfa committed:
revert init accessors changes + update tests

iMostfa commented:
Thanks @xedin for your review. 

i reverted `init accessors` changes here 
https://github.com/swiftlang/swift/pull/74462/commits/2510a9242d63980dd8a15dfba406e9c5305d4c24

and applied your comments here: 
https://github.com/swiftlang/swift/pull/74462/commits/30b4898de52a918eb3a4a5ec8b5e3353423a1b4c

i would really appreciate if we run tests on linux to check that everything is okay 👌🏻

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7870,7 +7867,7 @@ ERROR(noncopyable_types_cannot_be_resilient, none,
       (const ValueDecl *))
 ERROR(noncopyable_cannot_have_read_set_accessor,none,
-      "noncopyable %select{variable|subscript}0 cannot provide a read and set accessor",
+      "noncopyable %select{variable|subscript}0 cannot provide a read and set member",
```
@kavon Can we remove `member` here completely and use `cannot provided both a read and set`?

iMostfa replied:
it will be like this: 
382b6131e14
i applied the change, i can revert it if it's not the best option <3

AnthonyLatsis replied:
As an aside, we could improve this message by:
* Displaying the descriptive declaration kind (`const ValueDecl *` +  `%kindonly0`) instead of a selection.
* Adding "both" for stress.
* Surrounding the accessor keywords with apostrophes.

iMostfa replied:
maybe i'm doing something in correct, but trying @AnthonyLatsis leave me with these new messages, i think they are not clear (or incorrect?)

```swift

public subscript(x x: Int) -> NonCopyableS { // expected-error {{noncopyable subscript cannot provide a read and set member}}
                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                       /*suggested comment*/ -> noncopyable setter cannot provide both read and set `accessor`

```
and for variables, it would be: 



```swift
public var x: NonCopyableS { // expected-error {{noncopyable variable cannot provide a read and set member}}
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                   /*suggested comment*/ ->   noncopyable property cannot provide both read and set `accessor`




i find `subscript` one is less helpful ? i get that subscript is (setter/setting) by defination, but are we getting any value from replacing them with the `descriptiveDeclaration` ? 

i'm fine with adding `both` and `apostrophes ` btw

AnthonyLatsis replied:
A subscript can be read-only. You appear to be passing a setter instead of the subscript to the `diagnose` method.

xedin commented on test/Macros/accessor_macros.swift:
```diff
@@ -171,6 +171,6 @@ macro addGetterMacro() =
 struct S {
   @addGetterMacro let x: Int
-  // expected-warning@-1 {{cannot expand accessors on variable declared with 'let'; this is an error in the Swift 6 language mode}}
+  // expected-warning@-1 {{cannot expand macros on variable declared with 'let'; this is an error in the Swift 6 language mode}}
```
Test needs to be updated

iMostfa replied:
done: 
3d845f3e991

xedin reviewed (approved):
LGTM with a few minor changes still required.

iMostfa committed:
update acessor_macros tests

iMostfa committed:
remove member from noncopyable_cannot_have_read_set_accessor

AnthonyLatsis commented:
I think we are at a point where this is a documentation, not a compiler diagnostics QOI, problem. And I really do hate to barge in on a good first issue implementation with a note of disapproval, but I cannot help but worry about the turn in quality in many of these new phrasings. Removing the word "accessor" from diagnostic messages impedes clarity and accuracy far more often than it helps — I dare say, there are very few people in our community that know what a getter/setter is, write code that can hit these diagnostics, and would still be perplexed with what an accessor is. Plus, we now have accessor macros and `init` accessors to document in the book. There is no escaping the word "accessor" in the long run.

xedin commented:
The few diagnostics that are still modified by this PR don't regress quality in my opinion because they mostly clarify what kind of an accessor (which is a term of art here really) they refer to.

iMostfa committed:
update effectful_property_wrapper

iMostfa commented:
> we now have accessor macros and init accessors to document in the book
regarding this point, i think that macros exposes more complex terms due its nature and how it works (if i'm writing macros, i'm going to learn/discover more jargon related to the AST/compiler..etc) 

but as a regular swift user the word `accessor` is still not user-facing known word, at least until now.
so adding more context (to the kind of accessor) or replacing with better wordings shouldn't hurt (and here i'm speaking as a swift user)

AnthonyLatsis commented:
> i think that macros exposes more complex terms due its nature and how it works

Consider expression macros, where expressions are a fundamental feature independent of macros. Similarly, accessors by themselves are not an advanced feature. There is no reason to gate this word behind macros. Macros just happened to be the thing that caused us to finally expose the term in the book.

The macro role is called "accessor" not because macros are so advanced, for a different reason.

> but as a regular swift user the word `accessor` is still not a knowing user-facing word, at least until now.

Well, at least some diagnostic messages have had it for a long time. But yeah, my case is that if we feel uncomfortable about using this word in user-facing text at this point in time, then we should consider tending to the book instead. [We made it a user-facing term](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/attributes#attached), thus largely resolving #55887, but apparently forgot to define it and adjust the language reference.

> The few diagnostics that are still modified by this PR don't regress quality in my opinion because they mostly clarify what kind of an accessor (which is a term of art here really) they refer to.

We might do well to exercise more caution around messages that state rules. Specialization of a rule is an informational trade-off: you give away information about the general rule for information about how it looks like in a special case. This is understandably justified if the specialized rule is simpler or more convenient, that is, if the information you get in return is useful. If we are pointing to a getter anyway, what is the point in losing important information for a redundant "getter" in the message? Would you rather learn that `final` cannot be applied to this getter, or that it cannot be applied to accessors in general, including this getter?

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3975,8 +3975,8 @@ ERROR(attr_incompatible_with_objc,none,
 ERROR(final_not_on_accessors,none,
-      "'final' cannot be applied to accessors, it must be put on the "
-      "%select{var|let|subscript}0", (unsigned))
+      "'final' cannot be applied to %1, it must be put on the "
+      "%select{var|let|subscript}0", (unsigned, DescriptiveDeclKind))
```
I think this could be argued either way depending on a personal preference just like the rest of the points. If more than one accessor there declares ‘final’ we still get a diagnostic per use, so using a kind vs. ‘accessors’ seems like a moot point for clarity.

xedin replied:
In this spirit this particular diagnostic could be changed into something like - `only {property, subscript} itself can be marked as  ‘final’` and attach it to the parent declaration with note and fix-it per ‘final’ use.

AnthonyLatsis replied:
> so using a kind vs. ‘accessors’ seems like a moot point for clarity.

The stress is on the volume of educational information that we can offer without losing clarity, rather than clarity itself. If we can equip the user with the knowledge to not make a similar mistake, like putting `final` on a setter this time, as in `class can only inherit from another class` vs. `class cannot inherit from struct 'S'`, I cannot think why that wouldn’t be a better user experience.

But anyway, your suggestion sounds fine for a compromise 🙂 `'final' must be put on the {var, subscript} instead` is another possibility, but I think I like yours better.

xedin replied:
I understand what you are saying but when it comes to accessors at least there are ways to rephrase these messages that don’t involve terminology and don’t loose quality. 

When it comes to the message you referenced I think it could say `… cannot inherit from struct …` and put additional information into a note to expand what is possible if users need that.

xedin replied:
I think it all depends on what is really emphasized by a diagnostic…

AnthonyLatsis replied:
> When it comes to the message you referenced I think it could say `… cannot inherit from struct …` and put additional information into a note to expand what is possible

Mm, this did not occur to me. I love it!

iMostfa committed:
Add apostrophes around @objc

iMostfa committed:
Add property wrapper to a message

iMostfa committed:
Revert the removal of an accessor word

iMostfa commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7870,7 +7867,7 @@ ERROR(noncopyable_types_cannot_be_resilient, none,
       (const ValueDecl *))
 ERROR(noncopyable_cannot_have_read_set_accessor,none,
-      "noncopyable %select{variable|subscript}0 cannot provide a read and set accessor",
+      "noncopyable %select{variable|subscript}0 cannot provide a read and set",
```
reverted the change in: 
0851500a05471f2519391ce10982e8c65045cffd

iMostfa commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7204,8 +7201,8 @@ ERROR(property_wrapper_computed, none,
-    "property wrappers currently cannot define an 'async' or 'throws' accessor",
-    ())
+      "'wrappedValue' property cannot currently define an 'async' or 'throws' %0",
+    (DescriptiveDeclKind))
```
done in fef36928b9e02e7bbdfe3bc937cde68f68c4302d

iMostfa commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -6385,9 +6385,8 @@ NOTE(objc_witness_objc_requirement,none,
-
 ERROR(objc_observing_accessor, none,
-      "observing accessors are not allowed to be marked @objc", ())
+      "%0 is not allowed to be marked @objc", (DescriptiveDeclKind))
```
@AnthonyLatsis 
added `@objc`  here: 
ef9a3e62161b8061f2a6f2fdd684384728bef4d0

AnthonyLatsis replied:
Apostrophes (`'`), not backticks (`` ` ``).

iMostfa replied:
my bad, fixed c426560b3a4ce93d96ff4987614d0409ae4a5811

iMostfa committed:
reword final_not_on_accessors

iMostfa commented:
Hi @AnthonyLatsis and @xedin !

thank you so much for the constructive discussions about the messages, i enjoyed reading them.

i applied your feedback in https://github.com/swiftlang/swift/commit/ef9a3e62161b8061f2a6f2fdd684384728bef4d0, https://github.com/swiftlang/swift/commit/0851500a05471f2519391ce10982e8c65045cffd, and https://github.com/swiftlang/swift/commit/fef36928b9e02e7bbdfe3bc937cde68f68c4302d


finally for 

```c++
ERROR(final_not_on_accessors,none,
      "only %select{var|let|subscript}0 itself can be marked as 'final'"
```

i rephrased the message, yet i'm still missing the fix-it, would appreciate any hint about how to triggering this diagnostics message from the compiler, i tried the following but it didn't help: 

```swift
class MyClass {
     var myProperty: Int {
         final get {
          ^~~~~~~~  Cannot find 'final' in scope
            return 42
        }
    }
}
```

can i ask you please to run the tests on linux ? to be able to have an exhaustive list for falling tests !

xedin commented:
@swift-ci please smoke test Linux platform

iMostfa committed:
Fixed tests

iMostfa commented:
Hello @xedin  Thanks for running tests.

i fixed tests in https://github.com/swiftlang/swift/pull/74462/commits/81e98af527a65f2cbafbc983a137b3109d6a01fb

NOTE: there was a faulty merge commit that i did, because of it almost all Swift Members were assigned to my PR, my apologies !!
i reverted bad-changes, and re-opened the PR

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -6312,9 +6315,8 @@ NOTE(objc_witness_objc_requirement,none,
       "unable to resolve type for _opaqueReturnTypeOf attribute", ())
 
-
 ERROR(objc_observing_accessor, none,
-      "observing accessors are not allowed to be marked @objc", ())
+      "%0 is not allowed to be marked '@objc'", (DescriptiveDeclKind))
```
This is another instance of trading generality for but a slightly more precise description of what the diagnostic is already pointing at, which I believe is not a fair exchange. I can think of 3 ways to both retain generality and get rid of the "accessor" term.

1. Replace the template parameter with "observers", [as they are called in the book](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/declarations).
1. Attach a note articulating the general rule. This is mostly text duplication because the rule hasn’t gotten any simpler— far from ideal.
1. `observers (%0) are not allowed ...`.

AnthonyLatsis replied:
Thoughts?

iMostfa replied:
agreed, i applied third suggestion here: 
618051a3edf

i think it makes more sense to let the diagnostics message give more information in the case, because ad a swift-user i would get that it's about all observers

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -5786,6 +5786,9 @@ ERROR(distributed_property_cannot_be_static,none,
 ERROR(distributed_property_can_only_be_computed,none,
       "%kind0 cannot be 'distributed', only computed properties can",
       (const ValueDecl *))
+ERROR(distributed_property_accessor_only_get_can_be_distributed,none,
+      "only getters may be 'distributed'",
+      ())
```
This seems to be unused.

iMostfa replied:
done. [4ff3b9c](https://github.com/swiftlang/swift/pull/74462/commits/4ff3b9c8f03fecfa36f15ef2e436f3ad10697b70)

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3952,8 +3952,8 @@ ERROR(attr_incompatible_with_objc,none,
 ERROR(final_not_on_accessors,none,
-      "'final' cannot be applied to accessors, it must be put on the "
-      "%select{var|let|subscript}0", (unsigned))
+      "only %select{var|let|subscript}0 itself can be marked as 'final'"
+      ,(unsigned))
```
The are several issues with this diagnostic:
1. As you noted earlier, a parsing error is getting in its way.
1. Consequently, there are no tests for it.
1. A `let` cannot be `final`.

I will open an issue for 1 & 2, don’t bother addressing them in this PR. Suggestion for 3:
* ```suggestion
      "only the %select{property|subscript}0 itself can be marked as 'final'"
      ,(bool))
  ```
* Update emission site accordingly.

iMostfa replied:
done, but do we know why we previously supported let ? sounds odd to me
d00dd6ca27cd3de3e75245ef20a6d7eaa6565899

AnthonyLatsis replied:
> but do we know why we previously supported let 

We never did. The inclusion of `let` in the message was probably an oversight.

AnthonyLatsis replied:
> I will open an issue for 1 & 2

https://github.com/swiftlang/swift/issues/75933.

iMostfa committed:
remove distributed_property_accessor_only_get_can_be_distributed

iMostfa committed:
Remove 'currently' from property_wrapper_effectful

iMostfa commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7131,8 +7133,8 @@ ERROR(property_wrapper_computed, none,
-    "property wrappers currently cannot define an 'async' or 'throws' accessor",
-    ())
+      "property wrapper 'wrappedValue' property cannot currently define an 'async' or 'throws' %0",
+    (DescriptiveDeclKind))
```
personally i agree with you about this, because "yet" and "currently" doesn't imply that in future it's going to be supported. 

done in: 4d2c617976d

iMostfa committed:
Add 'observers' to objc_observing_accessor

iMostfa committed:
Remove let from final_not_on_accessors

kavon commented:
@swift-ci test

xedin commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1377,8 +1377,11 @@ void AttributeChecker::visitObjCAttr(ObjCAttr *attr) {
-      if (!accessor->isGetterOrSetter())
-        error = diag::objc_observing_accessor;
+      if (!accessor->isGetterOrSetter()) {
+        diagnoseAndRemoveAttr(attr, diag::objc_observing_accessor, accessor->getDescriptiveKind()).limitBehavior(behavior);
```
Nit: this needs formatting.

iMostfa replied:
Done here.
https://github.com/swiftlang/swift/pull/74462/commits/101c8074d0f1107766d7a50d7ffc4b74640ebdba

xedin commented on test/attr/attr_objc.swift:
```diff
@@ -142,10 +142,10 @@ class subject_getterSetter1 {
 
   var observingAccessorsVar1: Int = 0 {
-    @objc // expected-error {{observing accessors are not allowed to be marked @objc}} {{5-11=}}
+    @objc // expected-error {{willSet observer is not allowed to be marked @objc}} {{5-11=}}
```
Should `willSet` and `didSet` be in quotes?

iMostfa replied:
this was an outdated message

here we have the new message
https://github.com/swiftlang/swift/pull/74462/commits/3cd17c788695c79488d5bf8dc49528e2c53da9d3

> observers (didSet observer) are not allowed to be marked '@objc'


`didSet observer` is `accessor->getDescriptiveKind())`

xedin replied:
I think we should still mark that with quotes maybe?

AnthonyLatsis replied:
> I think we should still mark that with quotes maybe?

We should, but I think we can leave this out for someone else to fix as a first contribution. The missing quotes in the descriptive decl kind string is an issue of its own:

https://github.com/swiftlang/swift/blob/fa72680d28380f81a755a2975ba3afb0aaef06f9/lib/AST/Decl.cpp#L370-L371

@iMostfa Up to you.

xedin commented on lib/AST/Decl.cpp:
```diff
@@ -8357,8 +8357,9 @@ void VarDecl::emitLetToVarNoteIfSimple(DeclContext *UseDC) const {
+      auto descriptiveKindName = Decl::getDescriptiveKindName(FD->getDescriptiveKind());
       auto diags = d.diagnose(FD->getFuncLoc(), diag::change_to_mutating,
-                              isa<AccessorDecl>(FD));
+                              isa<AccessorDecl>(FD),descriptiveKindName);
```
Nit: Formatting

iMostfa replied:
done https://github.com/swiftlang/swift/pull/74462/commits/101c8074d0f1107766d7a50d7ffc4b74640ebdba

xedin commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2535,10 +2538,11 @@ void AttributeChecker::visitFinalAttr(FinalAttr *attr) {
-      diagnose(attr->getLocation(), diag::final_not_on_accessors, Kind)
+        isProperty = true;
+      diagnose(attr->getLocation(), diag::final_not_on_accessors,
+               isProperty)
```
Why cannot we just use `isa<VarDecl>(accessor->getStorage())` here directly instead of `isProperty`?

iMostfa replied:
i changed it to `isProperty = isa<VarDecl>(accessor->getStorage())`

https://github.com/swiftlang/swift/pull/74462/commits/101c8074d0f1107766d7a50d7ffc4b74640ebdba

xedin replied:
What I meant is:

```
 diagnose(attr->getLocation(), diag::final_not_on_accessors,
                 isa<VarDecl>(accessor->getStorage()))
```

And remove `isProperty`

iMostfa replied:
done here: 
https://github.com/swiftlang/swift/pull/74462/commits/ccacc70b02b899df1b5128a7c9d6c2a802e6797b

in my day to day job - iOS Programming-  i assume that `isProperty` will help the reader understand what's happening a little bit easier, and that the compiler will optimize it anyway, this's why i preferred it initially

xedin replied:
No worries, the operation itself should should give the read enough context to understand what is expected here without giving it a name.

xedin reviewed (commented):
I'm happy with the new phrasings, left a few minor inline comments. I think we can merge once these are addressed.

iMostfa committed:
Fix attr_objc

iMostfa committed:
Fix effectful_property_wrapper.swift

iMostfa committed:
Formatting

iMostfa commented:
Thanks @xedin i applied suggested changes, and fixed tests, we can run them again 🤞🏻 
let me know if there's anything is still missing :-)

xedin commented:
@swift-ci please test

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -6312,9 +6312,8 @@ NOTE(objc_witness_objc_requirement,none,
-
 ERROR(objc_observing_accessor, none,
-      "observing accessors are not allowed to be marked @objc", ())
+      "observers (%0) are not allowed to be marked '@objc'", (DescriptiveDeclKind))
```
```suggestion
      "observers (%kindonly0) are not allowed to be marked '@objc'", (AccessorDecl *))
```

This way you can pass the accessor decl itself at the emission site, without computing its descriptive decl kind. The diagnostic formatter will do this for us.

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -7131,8 +7130,8 @@ ERROR(property_wrapper_computed, none,
 ERROR(property_wrapper_effectful,none,
-    "property wrappers currently cannot define an 'async' or 'throws' accessor",
-    ())
+      "property wrapper's 'wrappedValue' property cannot define an 'async' or 'throws' %0",
+    (DescriptiveDeclKind))
```
```suggestion
      "property wrapper's 'wrappedValue' property cannot define an 'async' or 'throws' %kindonly0",
    (AccessorDecl *))
```

Ditto.

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -4985,8 +4985,8 @@ NOTE(candidate_is_not_assignable,none,
 NOTE(change_to_mutating,none,
-     "mark %select{method|accessor}0 'mutating' to make 'self' mutable",
-     (bool))
+     "mark %select{method|%1}0 'mutating' to make 'self' mutable",
+     (bool, StringRef))
```
Why pass a `StringRef` and not a `DescriptiveDeclKind`? Or better:

```suggestion
     "mark %select{method|%kindonly1}0 'mutating' to make 'self' mutable",
     (bool, FuncDecl *))
```

...and pass the function decl itself.

AnthonyLatsis reviewed (commented):
Thank you, this is looking much better. Just a few more tips and remarks.

xedin commented:
I think these could be addressed in a separate PR together with quotes.

iMostfa commented:
Hi @xedin , and @AnthonyLatsis 
1. Thanks a lot for the support and suggestions ❤️

2.i'm planning to address the next points, i just have some problems with building the compiler to be able to test, once i get them fixed, i will handle the points, along with my another PR that's still opened.
<img width="1567" alt="Screenshot 2024-08-17 at 08 59 03" src="https://github.com/user-attachments/assets/7d29edaf-23b0-4ee7-9985-43f349cdc371">

xedin commented:
Sounds good, thank you!

xedin commented:
You might want to run an update-checkout and try a clean build.

iMostfa commented:
@xedin i deleted everything and started from scratch. no luck
still trying to know what went wrong, and why, as i didn't do any changes to my environment, otherwise i will jump to the slack channel seeking help

iMostfa commented:
i was facing this issue. i opened a fix for it as i think it might affect others
https://github.com/swiftlang/swift/issues/75939
